src/vs/workbench/contrib/terminal/browser/terminalGroupService.ts

```diff
@@ -0,0 +1,316 @@
+import { Disposable } from 'vs/base/common/lifecycle';
+import { IContextKey, IContextKeyService } from 'vs/platform/contextkey/common/contextkey';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+import { IShellLaunchConfig } from 'vs/platform/terminal/common/terminal';
+import { ITerminalGroup, ITerminalGroupService, ITerminalInstance } from 'vs/workbench/contrib/terminal/browser/terminal';
+import { TerminalGroup } from 'vs/workbench/contrib/terminal/browser/terminalGroup';
+import { KEYBINDING_CONTEXT_TERMINAL_GROUP_COUNT } from 'vs/workbench/contrib/terminal/common/terminal';
+
+export class TerminalGroupService extends Disposable implements ITerminalGroupService {
+	declare _serviceBrand: undefined;
+
+	groups: ITerminalGroup[] = [];
+	activeGroupIndex: number = -1;
+	get instances(): ITerminalInstance[] {
+		return this.groups.reduce((p, c) => p.concat(c.terminalInstances), [] as ITerminalInstance[]);
+	}
+	activeInstanceIndex: number = -1;
+
+	private _terminalGroupCountContextKey: IContextKey<number>;
+	private _container: HTMLElement | undefined;
+
+	private readonly _onDidChangeActiveGroup = new Emitter<ITerminalGroup | undefined>();
+	get onDidChangeActiveGroup(): Event<ITerminalGroup | undefined> { return this._onDidChangeActiveGroup.event; }
+	private readonly _onDidDisposeGroup = new Emitter<ITerminalGroup>();
+	get onDidDisposeGroup(): Event<ITerminalGroup> { return this._onDidDisposeGroup.event; }
+	private readonly _onDidChangeGroups = new Emitter<void>();
+	get onDidChangeGroups(): Event<void> { return this._onDidChangeGroups.event; }
+
+	private readonly _onDidDisposeInstance = new Emitter<ITerminalInstance>();
+	get onDidDisposeInstance(): Event<ITerminalInstance> { return this._onDidDisposeInstance.event; }
+	private readonly _onDidChangeActiveInstance = new Emitter<ITerminalInstance | undefined>();
+	get onDidChangeActiveInstance(): Event<ITerminalInstance | undefined> { return this._onDidChangeActiveInstance.event; }
+	private readonly _onDidChangeInstances = new Emitter<void>();
+	get onDidChangeInstances(): Event<void> { return this._onDidChangeInstances.event; }
+
+	private readonly _onPanelOrientationChanged = new Emitter<Orientation>();
+	get onPanelOrientationChanged(): Event<Orientation> { return this._onPanelOrientationChanged.event; }
+
+	constructor(
+		@IContextKeyService private _contextKeyService: IContextKeyService,
+		@IInstantiationService private readonly _instantiationService: IInstantiationService
+	) {
+		super();
+
+		this.onDidDisposeGroup(group => this._removeGroup(group));
+
+		this._terminalGroupCountContextKey = KEYBINDING_CONTEXT_TERMINAL_GROUP_COUNT.bindTo(this._contextKeyService);
+		this.onDidChangeGroups(() => this._terminalGroupCountContextKey.set(this.groups.length));
+	}
+
+	get activeGroup(): ITerminalGroup | undefined {
+		if (this.activeGroupIndex < 0 || this.activeGroupIndex >= this.groups.length) {
```

Comment:
how could `activeGroupIndex` ever be greater than group length?

Reply:
I don't think it could unless something went wrong, that's just a sanity check to make sure we don't throw in that case.